docs/source/caching/programmatic-ids.mdx

```diff
@@ -0,0 +1,164 @@
+---
+title: Programmatic cache ids
+---
+
+> Using programmatic cache ids is the more flexible option but comes at the price of a complex API. For a compile-time, type-safe and simpler way to define your cache ids, you also use [declarative cache ids](declarative-ids).
+
+As explained in [the normalized cache page](normalized-cache), defining unique cache ids is important to avoid duplication and make sure your watchers get notified whenever a relevant change in the store happens.
+
+Cache ids are computed from two sources:
+
+1. From the object fields (`author.id`)
+2. From a field arguments (`author(id)`)
+
+
+1. is happening **after** a network request and is important to merge the network response with existing data. This is the most usual case.
+2. is happening **before** a network request and is important to avoid a network round trip if all the data is in the cache already. This is an optimization that can avoid some cache misses.
+
+With programmatic cache ids:
+
+1. uses `ObjectIdGenerator`
+2. uses `CacheKeyResolver`
+
+
+## ObjectIdGenerator
+
+To customize how cache ids are built from object fields, create a new instance of `ObjectIdGenerator`:
+
+```kotlin
+val objectIdGenerator = object : ObjectIdGenerator {
+  override fun cacheKeyForObject(obj: Map<String, Any?>, context: ObjectIdGeneratorContext): CacheKey? {
+    // Retrieve the id from the object itself
+    return CacheKey(obj["id"] as String)
+  }
+}
+```
+
+And pass it when configuring your normalized cache:
+
+```kotlin
+val apolloClient = ApolloClient.Builder()
+    .serverUrl("https://...")
+    .normalizedCache(
+        normalizedCacheFactory = cacheFactory,
+        objectIdGenerator = objectIdGenerator,
+    )
+    .build()
+```
+
+If you need the object typename, you can access it through `context`:
+
+```kotlin
+val objectIdGenerator = object : ObjectIdGenerator {
+  override fun cacheKeyForObject(obj: Map<String, Any?>, context: ObjectIdGeneratorContext): CacheKey? {
+    val typename = field.type.leafType().name
+    val id = obj["id"] as String
+
+    return CacheKey(typename, id)
+  }
+}
+```
+
+Note how this requires `id` to be present in the response. If a query does not return this field, the `CacheKey` will be inconsistent and duplication will happen.
+
+## CacheKeyResolver
+
+To customize how cache ids are built from fields arguments (before going to the network), create a new instance of `CacheKeyResolver`:
```

Comment:
We may want to clarify that the resolver must then be passed as an argument to `normalizedCache`

Reply:
100%, will do